Fix wrong scaling in Kelly error indicator

termi-official opened the pull request:
In case someone else also wondered why the Kelly error indicator required tolerances far below the ZZ estimators: It turns out the jump integrals in the Kelly error indicator are all scaled incorrectly in master. This PR corrects this initial mistake and adds a regression test for this behavior. Further the estimator tests are now extended to triangular and tetrahedral meshes. Also the performance has been improved by a few percent by hoisting some computations and reducing the total allocations.<!--GHEX{"author":"termi-official","assignment":"2023-06-04T20:54:50","editor":"tzanio","id":3693,"reviewers":["rw-anderson","TobiasDuswald"],"merge":"2023-09-10T20:47:29","approval":"2023-09-10T20:42:31"}XEHG-->
<!--GHEXTABLE-->
 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge|
 | --- | --- | --- | --- | --- | --- | --- |
| [#3693](https://github.com/mfem/mfem/pull/3693) | @termi-official | @tzanio | @rw-anderson + @TobiasDuswald | 6/4/23 | 9/10/23 | 9/10/23 | |
<!--ELBATXEHG-->
<details>
<summary>PR Checklist</summary>
    
- [ ] Code builds.
- [ ] Code passes `make style`.
- [ ] Update `CHANGELOG`:
    - [ ] Is this a new feature users need to be aware of? New or updated example or miniapp?
    - [ ] Does it make sense to create a new section in the `CHANGELOG` to group with other related features?
- [ ] Update `INSTALL`:
    - [ ] Had a new optional library been added? If so, what range of versions of this library are required? (*Make sure the external library is compatible with our BSD license, e.g. it is not licensed under GPL!*)
    - [ ] Have the version ranges for any required or optional libraries changed?
    - [ ] Does `make` or `cmake` have a new target?
    - [ ] Did the requirements or the installation process change? *(rare)*
- [ ] Update continuous integration server configurations if necessary (e.g. with new version requirements for each of MFEM's dependencies)
    - [ ] `.github`
    - [ ] `.appveyor.yml`
- [ ] Update `.gitignore`:
    - [ ] Check if `make distclean; git status` shows any files that were generated from the source by the project (not an IDE) but we don't want to track in the repository.
    - [ ] Add new patterns (just for the new files above) and re-run the above test.
- [ ] New examples:
    - [ ] All sample runs at the top of the example source file work.
    - [ ] Update `examples/makefile`:
      - [ ] Add the example code to the appropriate `SEQ_EXAMPLES` and `PAR_EXAMPLES` variables.
      - [ ] Add any files generated by it to the `clean` target.
      - [ ] Add the example binary and any files generated by it to the top-level `.gitignore` file.
    - [ ] Update `examples/CMakeLists.txt`:
      - [ ] Add the example code to the `ALL_EXE_SRCS` variable.
      - [ ] Make sure `THIS_TEST_OPTIONS` is set correctly for the new example.
   - [ ] List the new example in `doc/CodeDocumentation.dox`.
   - [ ] If new examples directory (e.g.`examples/pumi`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
      - [ ] Update or add example-specific documentation, see e.g. the `src/examples.md`.
      - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
      - [ ] In `examples.md`, list the example under the appropriate categories, add new categories if necessary.
      - [ ] Add a short description of the example in the "Extensive Examples" section of `features.md`.
- [ ] New miniapps:
   - [ ] All sample runs at the top of the miniapp source file work.
   - [ ] Update top-level `makefile` and `makefile` in corresponding miniapp directory.
   - [ ] Add the miniapp binary and any files generated by it to the top-level `.gitignore` file.
   - [ ] Update CMake build system:
      - [ ] Update the `CMakeLists.txt` file in the `miniapps` directory, if the new miniapp is in a new directory.
      - [ ] Add/update the `CMakeLists.txt` file in the new miniapp directory.
      - [ ] Consider adding a new test for the new miniapp.
   - [ ] List the new miniapp in `doc/CodeDocumentation.dox`
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), add it to `MINIAPP_SUBDIRS` in the `makefile`.
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
     - [ ] Update or add miniapp-specific documentation, see e.g. the `src/meshing.md` and `src/electromagnetics.md` files.
     - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
     - [ ] The miniapps go at the end of the page, and are usually listed only under a specific "Application (PDE)" category.
     - [ ] Add a short description of the miniapp in the "Extensive Examples" section of `features.md`.
- [ ] New capability:
   - [ ] All new public, protected, and private classes, methods, data members, and functions have full Doxygen-style documentation in source comments. Documentation should include descriptions of member data, function arguments and return values, template parameters, and prerequisites for calling new functions.
   - [ ] Pointer arguments and return values must specify whether ownership is being transferred or lent with the call.
   - [ ] Any new functions should include descriptions of their intended use e.g. for internal use only, user-facing, etc., along with references to example code whenever possible/appropriate.
   - [ ] Consider adding new sample runs in existing examples to highlight the new capability.
   - [ ] Consider saving cool simulation pictures with the new capability in the Confluence gallery (LLNL only) or submitting them, via pull request, to the gallery section of the `mfem/web` repo.
   - [ ] If this is a major new feature, consider mentioning it in the short summary inside `README` *(rare)*.
   - [ ] List major new classes in `doc/CodeDocumentation.dox` *(rare)*.
- [ ] Update this checklist, if the new pull request affects it.
- [ ] Run `make unittest` to make sure all unit tests pass.
- [ ] Run the tests in `tests/scripts`.
- [ ] (LLNL only) After merging:
   - [ ] Update internal tests to include the new features.
</details>

termi-official committed:
Stricter test coverage and regression test for Kelly checking exact integration of jump in abs function.

termi-official committed:
Fix scaling issue in Kelly error indicator.

termi-official committed:
Move allocating vectors out of the loops and delay transformation construction.

termi-official committed:
Apply style.

termi-official committed:
Partial rollback of LSZZ test extension.

termi-official commented:
cc @hughcars LSZZ fails with NC Tet (report https://github.com/mfem/mfem/actions/runs/5083398016/jobs/9134344629 in 082ad3323a5145f0730676ae94d72d2531cd983d).

termi-official committed:
Fix CI.

tzanio commented:
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

tzanio committed:
Merge branch 'master' into do/fix-kelly-scaling

TobiasDuswald commented on tests/unit/fem/test_estimator.cpp:
```diff
@@ -618,7 +655,7 @@ TEST_CASE("Kelly Error Estimator on 3D NCMesh",
          REQUIRE(local_errors(i) == MFEM_Approx(0.0));
       }
-      REQUIRE(estimator.GetTotalError() == MFEM_Approx(0.0));
+      // REQUIRE(estimator.GetTotalError() == MFEM_Approx(0.0));
```
@termi-official Is it intentional that the tests on the global error are commented out now? If I uncomment these lines (658, 677, 696), the unit tests fail. I think these lines should be removed if the test is no longer appropriate. However, if we can test the TotalError here similar to 718 / 740, this would be even better. As I see it, `Kelly Error Estimator on 3D NCMesh/Perfect Approximation XYZ" no longer tests for the total error. Can you please clarify?

termi-official replied:
I removed the tests (see my comment in the lines 275-283 in the implementation).

TobiasDuswald commented on fem/estimators.cpp:
```diff
@@ -324,6 +295,10 @@ void KellyErrorEstimator::ComputeEstimates()
       }
+      else     // We area t a boundary face
+      {
+         // TODO implement boundary flux difference
```
Also, do you intend to address thee TODOs you added before merging? (also in line 245, 347) ?

termi-official replied:
No, this PR is a bug-fix and I do not intend to add new features. However, I forgot to add the hints in the original PR, so I just wanted to leave them in this PR (since I felt like it is minor). I can remove the hints if you feel they are out of place.

TobiasDuswald replied:
Having hints is better than not having hints. For me, it's perfectly fine. Thanks for clarifying!

termi-official commented on fem/estimators.cpp:
```diff
@@ -214,108 +214,79 @@ void KellyErrorEstimator::ComputeEstimates()
-               IntegrationPoint ip;
-               FT->Loc2.Transform(fip, ip);
+               // Evaluate flux jump at IP on element 1
+               FT->Loc1.Transform(fip, ip);
+               flux->GetVectorValue(FT->Elem1No, ip, val);
+               double jump = val * normal;
 
-               Vector val(flux_space->GetVDim());
+               // Evaluate flux jump at IP on element 2
+               FT->Loc2.Transform(fip, ip);
                flux->GetVectorValue(FT->Elem2No, ip, val);
+               jump -= val * normal;
```
@TobiasDuswald interestingly, these lines already fail. `flux->GetVectorValue(FT->Elem1No, ip, val);` returns slightly different values than `flux->GetVectorValue(FT->Elem2No, ip, val);` for the corresponding transformed integration point. I think the transformation is slightly inexact resulting in the differences here. This leads to some accumulate for higher order elements because they have just more quadrature points per face to integrate properly.

hughcars replied:
I think I'm missing the context here: When you say higher order elements here, do you mean higher solution order, i.e. P, or higher geometry order i.e. curved elements and Q? How big is "slightly different"? The transformation matrices should be consistent for vectors within the tangent space defined by the face. Definitely for linear elements, (and imo for curved too, but I concede that's much harder to pull off).

termi-official committed:
Remove invalid test invariants.

TobiasDuswald reviewed (approved):
Alright, thanks for getting back to me regarding the previous points @termi-official. Other than that it looked good to me!

termi-official committed:
Fix bug.

termi-official committed:
Space.

termi-official commented:
Okay I found another bug which I could track down with ASAN.

```
AddressSanitizer:DEADLYSIGNAL
=================================================================
==1109908==ERROR: AddressSanitizer: SEGV on unknown address 0x000000000034 (pc 0x55672468f3c5 bp 0x7ffdd9cc45b0 sp 0x7ffdd9cc45a0 T0)
==1109908==The signal is caused by a READ memory access.
==1109908==Hint: address points to the zero page.
    #0 0x55672468f3c5 in mfem::FiniteElement::GetOrder() const /home/dogiermann/Repos/mfem-master/fem/fe/fe_base.hpp:326
    #1 0x556724fe9fe2 in mfem::FiniteElementSpace::GetFaceOrder(int, int) const /home/dogiermann/Repos/mfem-master/fem/fespace.cpp:2661
    #2 0x556724eedb7a in mfem::KellyErrorEstimator::ComputeEstimates() /home/dogiermann/Repos/mfem-master/fem/estimators.cpp:249
    #3 0x556724ef1b97 in mfem::KellyErrorEstimator::GetLocalErrors() /home/dogiermann/Repos/mfem-master/fem/estimators.hpp:656
    #4 0x556724847b21 in mfem::ThresholdRefiner::ApplyImpl(mfem::Mesh&) /home/dogiermann/Repos/mfem-master/mesh/mesh_operators.cpp:95
    #5 0x556724657c8e in mfem::MeshOperator::Apply(mfem::Mesh&) (/home/dogiermann/Repos/cycler/build/src/cycler-rkdg-amr+0x38cc8e) (BuildId: 1eab2866794aa2c1f5b2333c58421d03f9cdbe47)
```

This happens when `GetFaceOrder` is called on a finite element space with L2 elements. Since I am not sure how to handle the error cleanly I will leave it as subsequent work.

termi-official committed:
Style.

termi-official committed:
Add option to not use the flux coefficient.

termi-official committed:
Style.

termi-official committed:
Fix.

termi-official committed:
Remove dead docstring.

termi-official committed:
Revert last commit.

rw-anderson reviewed (approved):
Looks good to me. I wasn't sure about the comment regarding DG was the only thing.

termi-official commented on fem/estimators.cpp:
```diff
@@ -207,115 +211,88 @@ void KellyErrorEstimator::ComputeEstimates()
+            auto FT = mesh->GetFaceElementTransformations(fi);
+
+            // TODO Refactor this computation into a user-facing function.
+            // auto &int_rule = IntRules.Get(FT->FaceGeom, 2 * xfes->GetFaceOrder(fi)); // NOTE: This fails for DG
```
Thanks!

I am currently preparing a publication where I use Kelly for AMR with DG discretization and it works really better than expected for my problem, although I cannot give a general recommendation to always use Kelly. The problem here is that MFEM segfaults in the call to `xfes->GetFaceOrder(fi)` if xfes is $L_2$. From my understanding the segfault is triggered due to $L_2$ spaces not having an interpolation on faces defined. I think the current DG diffusion integrator has the same issue as this code - see https://github.com/mfem/mfem/blob/f97a770dcaa7498c93069ca7f27d25d9655acd86/fem/bilininteg.cpp#L3424-L3434 .

If you have a better solution in mind I am happy to implement it!

termi-official replied:
Thanks for taking time for the valuable comment @hughcars . Correct me if I am wrong, but the face order would be the interpolation order constrained to the face, right? Here, for mixed order H1 spaces it would be the minimal order of the space, because of the continuity constraint, however for mixed order L2 spaces it whould be the maximum order, or am I missing something?

hughcars replied:
Yes sorry I got it the wrong way round, in H1 there would be the NC constraints activated, so the effective order of the higher space would be reduced, whereas for L2 there would not be, so the selection would be the higher order.

hughcars commented on fem/estimators.cpp:
```diff
@@ -358,91 +339,67 @@ void KellyErrorEstimator::ComputeEstimates()
 
-      auto &int_rule = IntRules.Get(FT->FaceGeom, 2 * xfes->GetFaceOrder(0));
+      // TODO Refactor this computation into a user-facing function.
+      // auto &int_rule = IntRules.Get(FT->FaceGeom, 2 * xfes->GetFaceOrder(0)); // NOTE: This fails for DG
```
As above, I suggest fixing GetFaceOrder for DG spaces to resolve to the minimum or maximum of the attached face orders.

termi-official replied:
I will take a look, but I might need a few days.

termi-official replied:
I cannot reproduce the crash after merging master. In addition I added a test to catch regression on this. However, I do not have a query for the variable order case here, because I cannot find a good angle on fixing GetFaceOrder on the variable order case. This will also require additional test coverage. Probably this should be fixed in a separate PR, because this PR was merely proposed to fix the faulty scaling of the error.

hughcars commented on fem/estimators.cpp:
```diff
@@ -358,91 +339,67 @@ void KellyErrorEstimator::ComputeEstimates()
       for (int i = 0; i < nip; i++)
       {
-         // Evaluate flux vector at integration point
+         // Setup integration point
```
Setup is the noun, set up is the verb which I think is what you meant.

hughcars commented on fem/estimators.hpp:
```diff
@@ -598,6 +598,8 @@ class KellyErrorEstimator final : public ErrorEstimator
    flux_space; /**< @brief Ownership based on own_flux_fes. */
    bool own_flux_fespace; ///< Ownership flag for flux_space.
 
+   bool with_coef; ///< Whether the flux or gradient should be used.
```
This comment was a surprise, which suggests the name "with_coef" is too vague. Consider a better name, `using_flux` or some such?

termi-official replied:
I can do this, but want to note that I chose this name to be consistent with the remaining error estimators. Should I change the naming there too?

hughcars commented on tests/unit/fem/test_estimator.cpp:
```diff
@@ -60,7 +60,8 @@ TEST_CASE("Least-squares ZZ estimator on 2D NCMesh", "[NCMesh]")
 {
    // Setup
    const auto order = GENERATE(1, 3, 5);
```
Housekeeping in tests using `GENERATE`, it is helpful to include `CAPTURE(order);` as this will allow differentiation between failures for the parameterized tests, when looking at log files. Otherwise they're all conflated.

termi-official replied:
Thanks for this trick. I did not know this. I will fix this one asap.

hughcars commented on tests/unit/fem/test_estimator.cpp:
```diff
@@ -221,7 +223,9 @@ TEST_CASE("Least-squares ZZ estimator on 3D NCMesh", "[NCMesh]")
    // Setup
    const auto order = GENERATE(2, 3);
-   Mesh mesh = Mesh::MakeCartesian3D(2, 2, 2, Element::HEXAHEDRON);
+   // const auto element = GENERATE(Element::TETRAHEDRON, Element::HEXAHEDRON);
```
Delete the debris or reenable for all element types (preferable).

termi-official replied:
The estimator is broken for Tets, so I will leave only the Hex test for now.

hughcars commented on tests/unit/fem/test_estimator.cpp:
```diff
@@ -244,7 +248,7 @@ TEST_CASE("Least-squares ZZ estimator on 3D NCMesh", "[NCMesh]")
       {
          REQUIRE(local_errors(i) < 1e-10);
       }
-      REQUIRE(estimator.GetTotalError() < 1e-10);;
+      REQUIRE(estimator.GetTotalError() < 1e-10);
```
`REQUIRE` is for establishing a predicate of whether a test can meaningfully continue, to establish a test return, use `CHECK`.

hughcars commented on tests/unit/fem/test_estimator.cpp:
```diff
@@ -618,7 +655,6 @@ TEST_CASE("Kelly Error Estimator on 3D NCMesh",
       {
          REQUIRE(local_errors(i) == MFEM_Approx(0.0));
       }
-      REQUIRE(estimator.GetTotalError() == MFEM_Approx(0.0));
```
Why is the perfect approximation no longer giving zero error? If it's no longer perfect, these sections might need renaming?

termi-official replied:
It was also not exactly zero before, but something in the order 1e-17, because of the bug. Now it is in the order of 1e-14, which is above the error check threshold. I am pretty sure this is due to the slight numerical differences being added up.

hughcars commented on tests/unit/fem/test_estimator.cpp:
```diff
@@ -669,13 +703,16 @@ TEST_CASE("Kelly Error Estimator on 3D NCMesh",
       }
-      REQUIRE(estimator.GetTotalError() > 0.0);
+      // Jump is 2 by construction (so jump squared is constant 4) and we have 2 elements per face.
+      REQUIRE(estimator.GetTotalError()== MFEM_Approx(sqrt(4*2)));
```
As above, `CHECK`.

hughcars reviewed (commented):
Some minor style issues. I'm not au fait with the theory of Kelly estimators so can't comment much to their applicability to DG, but generally any estimator that works for CG will work for DG but just be deficient in failing to account for all sources of error, as DG introduces more deviation. Given the Kelly error estimator is deficient by design (not accounting for all sources of error), this argument doesn't really hold, so there shouldn't be any real worry in applying it to broader schemes.

tzanio commented:
@termi-official, did you address all of @hughcars's comments?

termi-official commented:
Sorry, not yet. I am still investigating, because I am not 100% confident yet, that the implementation is completely correct. The estimator degrades on unstructured meshes and I am still looking if it is an inherent problem with this estimator or still some lingering bug in the code.

termi-official committed:
Merge branch 'master' into do/fix-kelly-scaling

termi-official committed:
Bulk rename coef -> flux

termi-official committed:
Test housekeeping via CAPTURE

termi-official committed:
Remove debris.

termi-official committed:
Relax estimator tests from require to check.

termi-official committed:
Grammar

termi-official committed:
Add a crash test for DG spaces. Face order query seems to work on master?

tzanio commented:
Merged in `next` for testing...

v-dobrev commented:
@termi-official, there seems to a problem with the sample run
```console
ex15 -est 1 -e 0.0001 -no-vis
```
Here's the output:
```console
Options used:
   --mesh ../data/star-hilbert.mesh
   --problem 0
   --nfeatures 1
   --order 2
   --max-err 0.0001
   --hysteresis 0.15
   --ref-levels 0
   --nc-limit 3
   --t-final 1
   --estimator 1
   --no-visualization
   --no-visit-datafiles

Time 0

Refinement:
Iteration: 1, number of unknowns: 101
Iteration: 2, number of unknowns: 361
Iteration: 3, number of unknowns: 1361
Iteration: 4, number of unknowns: 5001
Iteration: 5, number of unknowns: 5851
Iteration: 6, number of unknowns: 7021
Iteration: 7, number of unknowns: 9201
Iteration: 8, number of unknowns: 15681
Iteration: 9, number of unknowns: 33021
Iteration: 10, number of unknowns: 56201
Iteration: 11, number of unknowns: 75801
Iteration: 12, number of unknowns: 103301
Iteration: 13, number of unknowns: 131375
Iteration: 14, number of unknowns: 160489
PCG: Number of iterations: 500
PCG: No convergence!
Iteration: 15, number of unknowns: 188903
PCG: Number of iterations: 500
PCG: No convergence!
Iteration: 16, number of unknowns: 215573
PCG: Number of iterations: 500
PCG: No convergence!
Iteration: 17, number of unknowns: 237579
PCG: Number of iterations: 500
PCG: No convergence!
Iteration: 18, number of unknowns: 255463
PCG: Number of iterations: 500
PCG: No convergence!
Iteration: 19, number of unknowns: 270933
PCG: Number of iterations: 500
PCG: No convergence!
Iteration: 20, number of unknowns: 284545
PCG: Number of iterations: 500
PCG: No convergence!
Iteration: 21, number of unknowns: 295385
PCG: Number of iterations: 500
PCG: No convergence!
Iteration: 22, number of unknowns: 304995
PCG: Number of iterations: 500
PCG: No convergence!
Iteration: 23, number of unknowns: 312728
PCG: Number of iterations: 500
PCG: No convergence!
Iteration: 24, number of unknowns: 323202
PCG: Number of iterations: 500
PCG: No convergence!
Iteration: 25, number of unknowns: 346069
PCG: Number of iterations: 500
PCG: No convergence!
Iteration: 26, number of unknowns: 908049
PCG: Number of iterations: 500
PCG: No convergence!
Iteration: 27, number of unknowns: 3378115
PCG: Number of iterations: 500
PCG: No convergence!
Iteration: 28, number of unknowns: 13005696
PCG: Number of iterations: 500
PCG: No convergence!
Iteration: 29, number of unknowns: 50988041
PCG: Number of iterations: 500
PCG: No convergence!
Iteration: 30, number of unknowns: 201932544
terminate called after throwing an instance of 'std::bad_array_new_length'
  what():  std::bad_array_new_length
```

v-dobrev commented:
Note: Other sample runs with `-est 1` also seem to have similar issues.

termi-official commented:
Thanks for reporting @v-dobrev .

The issue is basically that the tolerance is too low. In this PR the Kelly estimator scales differently than on master, namely here the error scales with $\det(J)$ while on master it scales with $det(J)^2$ (if I have not messed up the scaling again). This leads to an over-refinement of the elements and the elements get so tiny that we hit float point stability limits in the solve. This quadratic scaling in the element size worked sometimes better, which is the reason why I have not noticed the issue for a long time. If you have a suggestion on how to improve test coverage (or some better way to check if the indicator is now really correctly implemented) I am happy to implemented it. I have also noticed that the indicator seems to shrink slower than it should during the AMR procedure, but I cannot find a good angle what to test. 

On my machine e.g. `./ex15 -est 0 -no-vis` and `mpirun -np 4 ex15p -est 1 -e 1e-4 -no-vis` both work

termi-official committed:
Update example configurations for Kelly.

v-dobrev committed:
In ex15/ex15p, update the Kelly estimator sample runs to have
more reasonable run times.

v-dobrev commented:
Re-merged in `next` for testing.

v-dobrev committed:
In ex15, update the Kelly estimator sample runs to make them run faster

v-dobrev commented:
Re-merged in `next` for testing.

termi-official commented:
I will push another fix here soon. After going back to pen and paper I realized that I deleted the face coefficient from the evaluation. Might need a few more hours to see if I can come up with better testing. Sorry for this.

termi-official committed:
Add back face weight term and some minor fixes of possibly faulty initialized variables.

termi-official committed:
More possibly faulty initializations in nonlinear geometries.

termi-official committed:
Make the implementation less confusing.

termi-official committed:
Remove crashing test.

termi-official commented:
I am still having trouble to check if the implementation really works. I have now modified example 15 to manually investigate some analytical examples and the computed errors. The computed errors follow the wavefront. For the analytical solution (parabola) the indicator has a small error (about 1e-7), which is not super close to zero. I also can observe that the indicator starts to fail when the threshold gets to low. With this in mind I think this might be a precision issue, however my recent commits fail to address the just described failure of the indicator. Below the modified example for reproduction purposes.

<details>
<summary>Modified ex 15</summary>
```cpp

#include "mfem.hpp"
#include <fstream>
#include <iostream>

using namespace std;
using namespace mfem;

// Choices for the problem setup. Affect bdr_func and rhs_func.
int problem;
int nfeatures;

// Prescribed time-dependent boundary and right-hand side functions.
double bdr_func(const Vector &pt, double t);
double rhs_func(const Vector &pt, double t);

// Update the finite element space, interpolate the solution and perform
// parallel load balancing.
void UpdateProblem(Mesh &mesh, FiniteElementSpace &fespace,
                   GridFunction &x, BilinearForm &a, LinearForm &b, FiniteElementSpace* err_fes,
                   GridFunction &eest);


inline double deg2rad(double deg) {
    return deg * M_PI / 180.0;
}

void rotate30(const Vector& x, Vector& y)
{
   const auto theta = 30.0 * M_PI / 180.0;
   y(0) = cos(theta)*x(0)  - sin(theta)*x(1);
   y(1) = sin(theta)*x(0)  + cos(theta)*x(1);
}

int main(int argc, char *argv[])
{
   // 1. Parse command-line options.
   problem = 0;
   nfeatures = 1;
   const char *mesh_file = "../data/star-hilbert.mesh";
   int order = 2;
   double t_final = 1.0;
   double max_elem_error = 5.0e-3;
   double hysteresis = 0.15; // derefinement safety coefficient
   int ref_levels = 0;
   int nc_limit = 3;         // maximum level of hanging nodes
   bool visualization = true;
   bool paraview = false;
   int which_estimator = 0;
   bool rotate_by_30 = false;

   OptionsParser args(argc, argv);
   args.AddOption(&mesh_file, "-m", "--mesh",
                  "Mesh file to use.");
   args.AddOption(&problem, "-p", "--problem",
                  "Problem setup to use: 0 = spherical front, 1 = ball.");
   args.AddOption(&nfeatures, "-n", "--nfeatures",
                  "Number of solution features (fronts/balls).");
   args.AddOption(&order, "-o", "--order",
                  "Finite element order (polynomial degree).");
   args.AddOption(&max_elem_error, "-e", "--max-err",
                  "Maximum element error");
   args.AddOption(&hysteresis, "-y", "--hysteresis",
                  "Derefinement safety coefficient.");
   args.AddOption(&ref_levels, "-r", "--ref-levels",
                  "Number of initial uniform refinement levels.");
   args.AddOption(&nc_limit, "-l", "--nc-limit",
                  "Maximum level of hanging nodes.");
   args.AddOption(&t_final, "-tf", "--t-final",
                  "Final time; start time is 0.");
   args.AddOption(&which_estimator, "-est", "--estimator",
                  "Which estimator to use: "
                  "0 = ZZ, 1 = Kelly. Defaults to ZZ.");
   args.AddOption(&rotate_by_30, "-rotate-by-30", "--rotate-by-30", "-no-rotate-by-30",
                  "--no-rotate-by-30",
                  "Rotate the input mesh by 30° or not.");
   args.AddOption(&visualization, "-vis", "--visualization", "-no-vis",
                  "--no-visualization",
                  "Enable or disable GLVis visualization.");
   args.AddOption(&paraview, "-paraview", "--paraview-datafiles", "-no-paraview",
                  "--no-paraview-datafiles",
                  "Save data files for ParaView (paraview.llnl.gov) visualization.");
   args.Parse();
   if (!args.Good())
   {
      args.PrintUsage(cout);
      return 1;
   }
   args.PrintOptions(cout);

   // 2. Read the mesh from the given mesh file on all processors. We can handle
   //    triangular, quadrilateral, tetrahedral, hexahedral, surface and volume
   //    meshes with the same code.
   Mesh mesh(mesh_file, 1, 1);
   if(rotate_by_30) mesh.Transform(rotate30);
   int dim = mesh.Dimension();
   int sdim = mesh.SpaceDimension();

   // 3. Project a NURBS mesh to a piecewise-quadratic curved mesh. Make sure
   //    that the mesh is non-conforming if it has quads or hexes and refine it.
   if (mesh.NURBSext)
   {
      mesh.UniformRefinement();
      if (ref_levels > 0) { ref_levels--; }
      mesh.SetCurvature(2);
   }
   for (int l = 0; l < ref_levels; l++)
   {
      mesh.UniformRefinement();
      mesh.RandomRefinement(0.2);
   }
   mesh.EnsureNCMesh(true);
   // Make sure tet-only meshes are marked for local refinement.
   mesh.Finalize(true);

   // 4. All boundary attributes will be used for essential (Dirichlet) BC.
   MFEM_VERIFY(mesh.bdr_attributes.Size() > 0,
               "Boundary attributes required in the mesh.");
   Array<int> ess_bdr(mesh.bdr_attributes.Max());
   ess_bdr = 1;

   // 5. Define a finite element space on the mesh. The polynomial order is one
   //    (linear) by default, but this can be changed on the command line.
   H1_FECollection fec(order, dim);
   FiniteElementSpace fespace(&mesh, &fec);

   // 6. As in Example 1p, we set up bilinear and linear forms corresponding to
   //    the Laplace problem -\Delta u = 1. We don't assemble the discrete
   //    problem yet, this will be done in the inner loop.
   BilinearForm a(&fespace);
   LinearForm b(&fespace);

   ConstantCoefficient one(1.0);
   FunctionCoefficient bdr(bdr_func);
   FunctionCoefficient rhs(rhs_func);

   BilinearFormIntegrator *integ = new DiffusionIntegrator(one);
   a.AddDomainIntegrator(integ);
   b.AddDomainIntegrator(new DomainLFIntegrator(rhs));

   // 7. The solution vector x and the associated finite element grid function
   //    will be maintained over the AMR iterations.
   GridFunction x(&fespace);

   // 8. Connect to GLVis. Prepare for ParaView output.
   char vishost[] = "localhost";
   int  visport   = 19916;

   socketstream sout;
   if (visualization)
   {
      sout.open(vishost, visport);
      if (!sout)
      {
         cout << "Unable to connect to GLVis server at "
              << vishost << ':' << visport << endl;
         cout << "GLVis visualization disabled.\n";
         visualization = false;
      }
      sout.precision(8);
   }

   ParaViewDataCollection paraview_dc("Example15", &mesh);
   paraview_dc.RegisterField("solution", &x);
   int vis_cycle = 0;

   // 9. As in Example 6, we set up an estimator that will be used to obtain
   //    element error indicators. The integrator needs to provide the method
   //    ComputeElementFlux. The smoothed flux space is a vector valued H1 (ZZ)
   //    or L2 (Kelly) space here.
   L2_FECollection flux_fec(order, dim);
   ErrorEstimator* estimator{nullptr};

   L2_FECollection err_fec(0, dim);
   auto err_fes = new FiniteElementSpace(&mesh, &err_fec, 1);
   GridFunction eest(err_fes);
   paraview_dc.RegisterField("eest", &eest);

   switch (which_estimator)
   {
      case 1:
      {
         auto flux_fes = new FiniteElementSpace(&mesh, &flux_fec, sdim);
         estimator = new KellyErrorEstimator(*integ, x, flux_fes);
         break;
      }

      default:
         std::cout << "Unknown estimator. Falling back to ZZ." << std::endl;
      case 0:
      {
         auto flux_fes = new FiniteElementSpace(&mesh, &fec, sdim);
         estimator = new ZienkiewiczZhuEstimator(*integ, x, flux_fes);
         break;
      }
   }

   // 10. As in Example 6, we also need a refiner. This time the refinement
   //     strategy is based on a fixed threshold that is applied locally to each
   //     element. The global threshold is turned off by setting the total error
   //     fraction to zero. We also enforce a maximum refinement ratio between
   //     adjacent elements.
   ThresholdRefiner refiner(*estimator);
   refiner.SetTotalErrorFraction(0.0); // use purely local threshold
   refiner.SetLocalErrorGoal(max_elem_error);
   //    refiner.PreferConformingRefinement();
   refiner.SetNCLimit(nc_limit);

   // 11. A derefiner selects groups of elements that can be coarsened to form
   //     a larger element. A conservative enough threshold needs to be set to
   //     prevent derefining elements that would immediately be refined again.
   ThresholdDerefiner derefiner(*estimator);
   derefiner.SetThreshold(hysteresis * max_elem_error);
   derefiner.SetNCLimit(nc_limit);

   // 12. The outer time loop. In each iteration we update the right hand side,
   //     solve the problem on the current mesh, visualize the solution and
   //     refine the mesh as many times as necessary. Then we derefine any
   //     elements which have very small errors.
   x = 0.0;
   for (double time = 0.0; time < t_final + 1e-10; time += 0.01)
   {
      cout << "\nTime " << time << "\n\nRefinement:" << endl;

      // Set the current time in the coefficients.
      bdr.SetTime(time);
      rhs.SetTime(time);

      // Make sure errors will be recomputed in the following.
      refiner.Reset();
      derefiner.Reset();

      // 13. The inner refinement loop. At the end we want to have the current
      //     time step resolved to the prescribed tolerance in each element.
      for (int ref_it = 1; ; ref_it++)
      {
         cout << "Iteration: " << ref_it << ", number of unknowns: "
              << fespace.GetVSize() << endl;

         // 14. Recompute the field on the current mesh: assemble the stiffness
         //     matrix and the right-hand side.
         a.Assemble();
         b.Assemble();

         // 15. Project the exact solution to the essential boundary DOFs.
         x.ProjectBdrCoefficient(bdr, ess_bdr);

         // 16. Create and solve the linear system.
         Array<int> ess_tdof_list;
         fespace.GetEssentialTrueDofs(ess_bdr, ess_tdof_list);

         SparseMatrix A;
         Vector B, X;
         a.FormLinearSystem(ess_tdof_list, x, b, A, X, B);

#ifndef MFEM_USE_SUITESPARSE
         GSSmoother M(A);
         PCG(A, M, B, X, 0, 500, 1e-12, 0.0);
#else
         UMFPackSolver umf_solver;
         umf_solver.Control[UMFPACK_ORDERING] = UMFPACK_ORDERING_METIS;
         umf_solver.SetOperator(A);
         umf_solver.Mult(B, X);
#endif

         // 17. Extract the local solution on each processor.
         a.RecoverFEMSolution(X, b, x);

         eest = estimator->GetLocalErrors();

         // 18. Send the solution by socket to a GLVis server and optionally
         //     save it in ParaView format.
         if (visualization)
         {
            sout.precision(8);
            sout << "solution\n" << mesh << eest << flush;
         }
         if (paraview)
         {
            paraview_dc.SetCycle(vis_cycle++);
            paraview_dc.SetTime(time);
            paraview_dc.Save();
         }

         //  19. Apply the refiner on the mesh. The refiner calls the error
         //      estimator to obtain element errors, then it selects elements to
         //      be refined and finally it modifies the mesh. The Stop() method
         //      determines if all elements satisfy the local threshold.
         refiner.Apply(mesh);
         if (refiner.Stop())
         {
            break;
         }

         // 20. Update the space and interpolate the solution.
         UpdateProblem(mesh, fespace, x, a, b, err_fes, eest);
      }

      // 21. Use error estimates from the last inner iteration to check for
      //     possible derefinements. The derefiner works similarly as the
      //     refiner. The errors are not recomputed because the mesh did not
      //     change (and also the estimator was not Reset() at this time).
      if (derefiner.Apply(mesh))
      {
         cout << "\nDerefined elements." << endl;

         // 22. Update the space and interpolate the solution.
         UpdateProblem(mesh, fespace, x, a, b, err_fes, eest);
      }

      a.Update();
      b.Update();
   }

   delete estimator;

   return 0;
}


void UpdateProblem(Mesh &mesh, FiniteElementSpace &fespace,
                   GridFunction &x, BilinearForm &a, LinearForm &b, FiniteElementSpace* err_fes,
                   GridFunction &eest)
{
   // Update the space: recalculate the number of DOFs and construct a matrix
   // that will adjust any GridFunctions to the new mesh state.
   fespace.Update();

   // Interpolate the solution on the new mesh by applying the transformation
   // matrix computed in the finite element space. Multiple GridFunctions could
   // be updated here.
   x.Update();

   // Free any transformation matrices to save memory.
   fespace.UpdatesFinished();

   err_fes->Update();
   eest.Update();
   err_fes->UpdatesFinished();

   // Inform the linear and bilinear forms that the space has changed.
   a.Update();
   b.Update();
}


const double alpha = 0.02;

// Spherical front with a Gaussian cross section and radius t
double front(double x, double y, double z, double t, int)
{
   double r = sqrt(x*x + y*y + z*z);
   return exp(-0.5*pow((r - t)/alpha, 2));
}

double front_laplace(double x, double y, double z, double t, int dim)
{
   double x2 = x*x, y2 = y*y, z2 = z*z, t2 = t*t;
   double r = sqrt(x2 + y2 + z2);
   double a2 = alpha*alpha, a4 = a2*a2;
   return -exp(-0.5*pow((r - t)/alpha, 2)) / a4 *
          (-2*t*(x2 + y2 + z2 - (dim-1)*a2/2)/r + x2 + y2 + z2 + t2 - dim*a2);
}

// Smooth spherical step function with radius t
double ball(double x, double y, double z, double t, int)
{
   double r = sqrt(x*x + y*y + z*z);
   return -atan(2*(r - t)/alpha);
}

double ball_laplace(double x, double y, double z, double t, int dim)
{
   double x2 = x*x, y2 = y*y, z2 = z*z, t2 = 4*t*t;
   double r = sqrt(x2 + y2 + z2);
   double a2 = alpha*alpha;
   double den = pow(-a2 - 4*(x2 + y2 + z2 - 2*r*t) - t2, 2.0);
   return (dim == 2) ? 2*alpha*(a2 + t2 - 4*x2 - 4*y2)/r/den
          /*      */ : 4*alpha*(a2 + t2 - 4*r*t)/r/den;
}

double parabola(double x, double y, double z, double t, int dim)
{
   return t*(x*x - y*y);
}

double parabola_laplace(double x, double y, double z, double t, int dim)
{
   return 0.0;
}

// Composes several features into one function
template<typename F0, typename F1, typename F2>
double composite_func(const Vector &pt, double t, F0 f0, F1 f1, F2 f2)
{
   int dim = pt.Size();
   double x = pt(0), y = pt(1), z = 0.0;
   if (dim == 3) { z = pt(2); }

   if (problem == 0)
   {
      if (nfeatures <= 1)
      {
         return f0(x, y, z, t, dim);
      }
      else
      {
         double sum = 0.0;
         for (int i = 0; i < nfeatures; i++)
         {
            double x0 = 0.5*cos(2*M_PI * i / nfeatures);
            double y0 = 0.5*sin(2*M_PI * i / nfeatures);
            sum += f0(x - x0, y - y0, z, t, dim);
         }
         return sum;
      }
   }
   else if (problem == 1)
   {
      double sum = 0.0;
      for (int i = 0; i < nfeatures; i++)
      {
         double x0 = 0.5*cos(2*M_PI * i / nfeatures + M_PI*t);
         double y0 = 0.5*sin(2*M_PI * i / nfeatures + M_PI*t);
         sum += f1(x - x0, y - y0, z, 0.25, dim);
      }
      return sum;
   }
   else
   {
      return f2(x, y, z, t, dim);
   }
}

// Exact solution, used for the Dirichlet BC.
double bdr_func(const Vector &pt, double t)
{
   return composite_func(pt, t, front, ball, parabola);
}

// Laplace of the exact solution, used for the right hand side.
double rhs_func(const Vector &pt, double t)
{
   return composite_func(pt, t, front_laplace, ball_laplace, parabola_laplace);
}
```
</details>

I also removed the crash test for the L2 spaces, because it turned out to not trigger the actual segfault. The "fixed" test which reproduces the crash has been moved to the issue linked above.

tzanio commented:
Given the above comment and the fact that the mfem-4.6 release is imminent, lets postpone this PR for after the release.

We can re-merge it in the new `next` branch to maintain its `in-review` status.

termi-official commented:
Hi @tzanio . Unfortunately nothing major. I am trying to construct another analytical test case where I want to show that the estimated error stays below the bound, because the refinement still looks kida off in tests. I try to circle back to this in 2 weeks, if that is okay. Semester started here and I have to prepare teaching, unfortunately.

adam-sim-dev commented:
What's the state of this PR?

termi-official commented:
I *think* it works, but I really stuggle with finding tests on this to verify.

adam-sim-dev commented on fem/estimators.cpp:
```diff
@@ -207,123 +211,106 @@ void KellyErrorEstimator::ComputeEstimates()
                {
+                  // This computes a weighted normal, so we divide by the weight to get back the normal.
                   CalcOrtho(FT->Face->Jacobian(), normal);
+                  normal /= FT->Face->Weight();
```
I agree that the normal of the face should be divided by the Weight, to get a unit normal vector.

adam-sim-dev commented:
> I *think* it works, but I really stuggle with finding tests on this to verify.

Could you please resolve the conflict first?

termi-official commented:
> > I _think_ it works, but I really stuggle with finding tests on this to verify.
> 
> Could you please resolve the conflict first? I am implementing my error estimator and using this PR as a reference.

Until when do you need it? I am unfortunately super occupied by some deadlines on my side atm.

adam-sim-dev commented:
> > > I _think_ it works, but I really stuggle with finding tests on this to verify.
> > 
> > Could you please resolve the conflict first? I am implementing my error estimator and using this PR as a reference.
> 
> Until when do you need it? I am unfortunately super occupied by some deadlines on my side atm.

No hurry. I can still refer to this branch. What I think is, once the conflicts are resolved, this PR can be reviewed by other developers. And I want to make sure the jump integrals are calculated correctly (See my comments above).

tzanio commented:
ping: @termi-official

termi-official commented:
Sorry for the delay here, but I still have not found a good way on how to test this, because I cannot come up with a minimal example that fails for the implementation on master. Maybe, if you have some time in one of your internal meetings, could you discuss how a useful invariant to test for the correctness of an error indicator could be? Right now the test is essentially looking whether the integrator is not fundamentally broken, i.e. positive and convergent.

termi-official committed:
Merge main

hughcars commented:
> I think this is considered breaking for some users, as the AMR process is direcly affected (due to changes in the tolerance). I also still have no clue on how to verify whether an error estimator is correctly implemented, because the one on master also looks like it works. How should we proceed here?

There's two mathematical properties a good estimator has, both of which are fiddly to test for in practice: reliability and efficiency. The best way to test those is a manufactured solution. You can compute the true error on a uniform refinement study with rhs that gives a known solution, then compare your estimated error against that true error computed elementwise. If you have a fixed constant factor *global* upper bound, it's reliable, if you have a fixed constant *local* lower bound, it's efficient. Plotting the min and max of the element local quantities as a function of `h` is the most principled way to demonstrate this data. Reliable means you can use it to drive AMR to completion, efficient means it will give you a reasonable mesh as it does so.

Looking at the [deal.II source comments](https://docs.mfem.org/4.8/classmfem_1_1KellyErrorEstimator.html#details), the Kelly estimator looks like it probably has a fairly weak equivalence relation with a standard strong form residual estimator, so should have both of those, but it will fail to account for any data oscillation, lacking any forcing function, source and misses any higher order terms from the hessian and non-diffusive terms. Thus the constants in those bounds are going to be pretty asymptotic rather than strict, as the `h_e` and `h_f` in the same term also indicates some inverse estimate was used in the construction, weakening the bounds further.

Not sure how practical this is for this PR, but thought a little context on why verifying (not even validating!) estimation can be fiddly, might be of interest.
